embassy-rp/src/dma.rs

```diff
@@ -1,82 +1,224 @@
+#[interrupt]
+unsafe fn DMA_IRQ_0() {
+    let ints0 = pac::DMA.ints0().read().ints0();
+
+    critical_section::with(|_| {
+        for channel in 0..CHANNEL_COUNT {
+            if ints0 & (1 << channel) == (1 << channel) {
+                CHANNEL_WAKERS[channel].wake();
+            }
+        }
+        pac::DMA.ints0().write(|w| w.set_ints0(ints0));
+    });
+}
+
+pub fn read<'a, C: Channel, W: Word>(ch: impl Peripheral<P = C> + 'a, from: *const W, to: &mut [W]) -> Transfer<'a, C> {
+    let (ptr, len) = crate::dma::slice_ptr_parts_mut(to);
+    copy_inner(ch, from as *const u32, ptr as *mut u32, len, W::size(), false, true)
+}
+
+pub fn write<'a, C: Channel, W: Word>(ch: impl Peripheral<P = C> + 'a, from: &[W], to: *mut W) -> Transfer<'a, C> {
+    let (from_ptr, len) = crate::dma::slice_ptr_parts(from);
+    copy_inner(ch, from_ptr as *const u32, to as *mut u32, len, W::size(), true, false)
+}
+
+pub fn copy<'a, C: Channel, W: Word>(ch: impl Peripheral<P = C> + 'a, from: &[W], to: &mut [W]) -> Transfer<'a, C> {
+    let (from_ptr, from_len) = crate::dma::slice_ptr_parts(from);
+    let (to_ptr, to_len) = crate::dma::slice_ptr_parts_mut(to);
+    assert_eq!(from_len, to_len);
+    copy_inner(
+        ch,
+        from_ptr as *const u32,
+        to_ptr as *mut u32,
+        from_len,
+        W::size(),
+        true,
+        true,
+    )
 }
 
-impl<T: Channel> Dma<T> {
-    pub fn copy(inner: T, from: &[u32], to: &mut [u32]) {
-        assert!(from.len() == to.len());
+fn copy_inner<'a, C: Channel>(
+    ch: impl Peripheral<P = C> + 'a,
+    from: *const u32,
+    to: *mut u32,
+    len: usize,
+    data_size: DataSize,
+    incr_read: bool,
+    incr_write: bool,
+) -> Transfer<'a, C> {
+    into_ref!(ch);
+
+    unsafe {
+        let p = ch.regs();
+
+        p.read_addr().write_value(from as u32);
+        p.write_addr().write_value(to as u32);
+        p.trans_count().write_value(len as u32);
+
+        compiler_fence(Ordering::SeqCst);
+
+        p.ctrl_trig().write(|w| {
+            w.set_data_size(data_size);
+            w.set_incr_read(incr_read);
+            w.set_incr_write(incr_write);
+            w.set_chain_to(ch.number());
+            w.set_en(true);
+        });
+
+        compiler_fence(Ordering::SeqCst);
+    }
+    Transfer::new(ch)
+}
+
+pub struct Transfer<'a, C: Channel> {
+    channel: PeripheralRef<'a, C>,
+}
+
+impl<'a, C: Channel> Transfer<'a, C> {
+    pub(crate) fn new(channel: impl Peripheral<P = C> + 'a) -> Self {
+        into_ref!(channel);
 
         unsafe {
-            let p = inner.regs();
+            let irq = interrupt::DMA_IRQ_0::steal();
+            irq.disable();
+            irq.set_priority(interrupt::Priority::P6);
 
-            p.read_addr().write_value(from.as_ptr() as u32);
-            p.write_addr().write_value(to.as_mut_ptr() as u32);
-            p.trans_count().write_value(from.len() as u32);
+            pac::DMA.inte0().write(|w| w.set_inte0(1 << channel.number()));
```

Comment:
this will disable irqs for all the other channels. Either do a `.modify` (within a CS to avoid races) or use the fancy "atomic set bit" register alias https://github.com/rp-rs/rp-hal/blob/main/rp2040-hal/src/atomic_register_access.rs#L21

EDIT: Actually I think you can simply enable all channel's irqs on boot.